scylla-proxy/src/proxy.rs

```diff
@@ -66,6 +66,100 @@ impl Node {
             response_rules,
         }
     }
+
+    pub fn builder() -> NodeWithoutRealAddress {
```

Comment:
I admit: I've chosen the more complex way to do that, for the sake of learning the Rust type system better and using [the typestate pattern](http://cliffle.com/blog/rust-typestate).
The boilerplate is separated from the `meaningful` code, so its presence should not be that intrusive.

Reply:
> I admit: I've chosen the more complex way to do that, for the sake of learning the Rust type system better and using [the typestate pattern](http://cliffle.com/blog/rust-typestate).

I understand that the typestate pattern is interesting, but I don't think it is a good fit here - or, at least, in its current implementation.

> The boilerplate is separated from the `meaningful` code, so its presence should not be that intrusive.

Using the builder looks actually pretty nice, and it's not a problem. The problem is the builder itself. It does something very simple but at the same time is very verbose. Adding another required field to the Node essentially requires you to copy-paste the entire struct and introduce a new state in the state machine.

I'm not sure that the added compile-time safety is worth the complexity. It could be worth it if you had some complex logic that uses the builder and it would be hard to validate that all code paths always use the builder correctly. However, this is only for tests - our tests usually have very simple control flow, are short and running them exercises all of the code paths. I think validating the config in runtime will be sufficient.